### PR TITLE
twap at height NTRN-357

### PR DESCRIPTION
**Pull request by quasisamurai**
[TASK](https://p2pvalidator.atlassian.net/browse/NTRN-357)

This is a patched `astroport oracle` contract with necessary `astroport contracts` and also some modifications 4 Neutron needs:

- Implemented twap_at_height query
- `PERIOD` const is now parameter
- implemented simple and complex tests 

since the astroport contract has nothing to integrate with at the moment, the tests are implemented directly in the repository (see integration.rs ) discussed w @swelf19 

How to test: you don't need to since you're seeing green check mark in this PR

### CODE REVIEW

File: contracts/astroport/oracle/src/contract.rs [hunk 1]
```diff
@@ -0,0 +1,261 @@
...
+                ))
+            }
+        })
+        .collect::<Result<Vec<(AssetInfo, Uint256)>, StdError>>()
+}
+
+/// Returns token TWAP value for given height.
+/// * **token** token for which we multiply its TWAP value by an amount.
+///
+/// * **height** amount of tokens we multiply the TWAP by.
```
Comment (pr0n00gler) [inline contracts/astroport/oracle/src/contract.rs:210]: Wrong comment

File: contracts/astroport/oracle/src/state.rs [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+use astroport::asset::{AssetInfo, PairInfo};
+use cosmwasm_std::{Addr, Decimal256, Uint128};
+use cw_storage_plus::{Item, SnapshotItem, Strategy};
+
+/// Stores the contract config at the given key
+pub const CONFIG: Item<Config> = Item::new("config");
+/// Stores the latest cumulative and average prices at the given key
+/// TODO: proper checkpoints
+pub const PRICE_LAST: SnapshotItem<PriceCumulativeLast> =
+    SnapshotItem::new("price_last", "", "", Strategy::EveryBlock);
```
Comment (pr0n00gler) [inline contracts/astroport/oracle/src/state.rs:12]:
```suggestion
    SnapshotItem::new("price_last", "price_last_checkpoints", "price_last_changelog", Strategy::EveryBlock);
```

**Review by pr0n00gler — CHANGES_REQUESTED** [review]

**Review by pr0n00gler — APPROVED** [review]

**Review by pr0n00gler — APPROVED** [review]

**Review by foxpy — APPROVED** [review]
code looks fine, though I am not sure how are we going to resolve dependency hell with cosmwasm-std and cw-X libraries.

**Review by pr0n00gler — APPROVED** [review]

**Review by oldremez — COMMENTED** [review]
I believe there's a lack of tests with a height between two snapshots. Current tests cover only cases with exactly those blocks at which the oracle was updated. But in a real situation, the queries will be done to the arbitrary block that in general lies between snapshots. And what is the expected behavior here: do we need to return the past TWAP calculated or the next one (since the block we get a query for lies in the period of time for which the next TWAP is calculated, right?).

Also, out of scope but maybe we should have a todo for it: if we keep the history anyway, shouldn't we allow updating at arbitrary blocks? I mean for any block we can find the previous snapshot that is more than the `update period` far away. Or it will lead to more gas consumption?

**Comment by quasisamurai** [comment]
plz don't merge this pr until this comment is removed

**Review by oldremez — APPROVED** [review]

**Pull request merged**